### PR TITLE
#164489890 Create and Read articles

### PR DESCRIPTION
**Pull request by itsgratien**
#### What does this PR do?
This PR allows users to create and read created articles
#### Description of Task to be completed?
- I designed a user interface for creating new articles
- I implemented editor which allow a user to write a story
- I created a component called CreateArticle.
- I created a route for navigating the user to create a new story called [localhost:3000/story/new-story](htttp://localhost:3000/story/new-story).
This PR has the following routes
 - [localhost:3000/story/new-story](http://localhost:3000/story/new-story)
 - [localhost:3000/story/45](http://localhost:3000/story/:id)

#### How should this be manually tested?
- Once you clone project
- Open project directory in your terminal
- write npm install to install package
- npm run dev to run a development server
- on your browser URL, write  [localhost:3000/story/new-story](htttp://localhost:3000/story/new-story)

#### Story with content, image,tags and title
Screen Shot 2019-05-26 at 14 34 26

##### When a user is going to create article
Screen Shot 2019-05-26 at 14 35 55

#####
Screen Shot 2019-05-24 at 13 23 10

#### What are the relevant pivotal tracker stories?
[164489890](https://www.pivotaltracker.com/story/show/164489890)

### CODE REVIEW

File: src/actions/article.js [hunk 1]
```diff
@@ -1,6 +1,17 @@
...
+  GET_SINGLE_ARTICLE
+} from './type';
 
+const token = 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpZCI6MTYsImVtYWlsIjoiZ3JhY2lhbjIwMjBAZ21haWwuY29tIiwiaXNBZG1pbiI6ZmFsc2UsImlhdCI6MTU1ODYxNTczNCwiZXhwIjoxNTYyOTM1NzM0fQ.zRM27LSAYifXieVs06dG6S3HaJPJjwnP6wA7WpFNKJA';
+const config = {
+  headers: {
+    'Content-Type': 'multipart/form-data',
+    Authorization: `Bearer ${token}`
+  }
+};
```
Comment (nakiwuge) [inline src/actions/article.js]: create a config file for this

File: src/actions/article.js [hunk 2]
```diff
@@ -18,7 +29,67 @@ const getAllArticle = () => (dispatch) => {
...
+  await dispatch(loading());
+  try {
+    const send = await axios.post(url, formData, config);
+    await dispatch({
+      type: CREATE_ARTICLE,
+      payload: send.data
+    });
+  } catch (error) {
+    await dispatch({
+      type: ARTICLE_ERROR,
```
Comment (nakiwuge) [inline src/actions/article.js]: Rename this type to CREATE_ARTICLE_FAILURE

File: src/actions/article.js [hunk 3]
```diff
@@ -18,7 +29,67 @@ const getAllArticle = () => (dispatch) => {
...
+  formData.append('title', data.title);
+  formData.append('body', data.body);
+  formData.append('tag', arr);
+  await dispatch(loading());
+  try {
+    const send = await axios.post(url, formData, config);
+    await dispatch({
+      type: CREATE_ARTICLE,
+      payload: send.data
+    });
```
Comment (nakiwuge) [inline src/actions/article.js]: create a function for ``` {
      type: CREATE_ARTICLE,
      payload: send.data
    } ``` and call it in the dispatch

File: src/components/articles/CreateArticle.js [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    this.state = {
+      title: '',
+      image: '',
+      tag: '',
+      loading: false,
+      open: false
+    };
+    this.BodyContent = React.createRef();
+  }
+
```
Comment (nakiwuge) [inline src/components/articles/CreateArticle.js:27]: You do not need a constructor when using arrow functions.  Just declare the state.

File: src/components/articles/CreateArticle.js [hunk 5]
```diff
@@ -0,0 +1,197 @@
...
+  onChange= (e) => {
+    this.setState({ [e.target.name]: e.target.value });
+  }
+
+  onSubmit=(e) => {
+    e.preventDefault();
+    // @set loading to true
```
Comment (nakiwuge) [inline src/components/articles/CreateArticle.js]: This comment is unnecessary

File: src/actions/article.js (hunk 2 above)
Comment (itsgratien) [reply src/actions/article.js]: thank you @zellymiriam

File: src/actions/article.js [hunk 6]
```diff
@@ -18,7 +29,67 @@ const getAllArticle = () => (dispatch) => {
...
+  const url = '/api/articles';
+  const arr = data.tag ? data.tag.join(',') : '';
+  const formData = new FormData();
+  formData.append('image', data.image);
+  formData.append('title', data.title);
+  formData.append('body', data.body);
+  formData.append('tag', arr);
+  await dispatch(loading());
+  try {
+    const send = await axios.post(url, formData, config);
```
Comment (nakiwuge) [inline src/actions/article.js]: please get this variable a better name

File: src/actions/article.js [hunk 7]
```diff
@@ -18,7 +29,67 @@ const getAllArticle = () => (dispatch) => {
...
+  dispatch({
+    type: REMOVE_TAG,
+    payload: data
+  });
+};
+// @single article
+const singleArticle = handle => async (dispatch) => {
+  const url = `/api/articles/${handle}`;
+  try {
+    const find = await axios.get(url, config);
```
Comment (nakiwuge) [inline src/actions/article.js]: Rename this variable to getArticle

File: src/actions/article.js [hunk 8]
```diff
@@ -18,7 +29,67 @@ const getAllArticle = () => (dispatch) => {
...
+  } catch (error) {
+    await dispatch({
+      type: ARTICLE_ERROR,
+      payload: error.response
+    });
+  }
+};
 
+// @ add tag
+const addTag = data => (dispatch) => {
```
Comment (zellymiriam) [inline src/actions/article.js]: remove this function and put the functionality in the component

File: src/actions/article.js [hunk 9]
```diff
@@ -18,7 +29,67 @@ const getAllArticle = () => (dispatch) => {
...
+  dispatch({
+    type: ADD_TAG,
+    payload: data
+  });
+};
+const removeTag = data => (dispatch) => {
```
Comment (zellymiriam) [inline src/actions/article.js]: remove this function and put the functionality in the component

File: src/components/articles/CreateArticle.js [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+setBold=() => {
+  document.execCommand('Bold', false, null);
+}
+
+openTag=(e) => {
+  e.preventDefault();
+  const { open } = this.state;
+  this.setState({ open: !open });
+}
```
Comment (nakiwuge) [inline src/components/articles/CreateArticle.js:66]: Use previous state for this toggle

File: src/tests/component/articles/Article.test.js [hunk 11]
```diff
@@ -0,0 +1,72 @@
+import React from 'react';
+import { shallow } from 'enzyme';
+import { CreateArticle } from '../../../components/articles/CreateArticle';
+
+
```
Comment (nakiwuge) [inline src/tests/component/articles/Article.test.js]: remove line

File: src/tests/component/articles/Article.test.js [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+  });
+  test('should open tag input', () => {
+    const spy = spyFunc(component.instance(), 'openTag');
+    const fakeEvent = { preventDefault: () => {} };
+    const input = component.find('[data-test="G-openTag"]');
+    input.simulate('click', fakeEvent);
+    expect(input.length).toBe(1);
+    expect(spy).toHaveBeenCalled();
+  });
+});
```
Comment (nakiwuge) [inline src/tests/component/articles/Article.test.js]: add a new line every after each test block and after declaring variables to improve readability

File: src/tests/actions/article.test.js [hunk 13]
```diff
@@ -0,0 +1,56 @@
+import moxios from 'moxios';
+import configuration from 'redux-mock-store';
+import ReduxThunk from 'redux-thunk';
+// @load action
+import getAllArticle, { createArticle, addTag, removeTag } from '../../actions/article';
+
```
Comment (nakiwuge) [inline src/tests/actions/article.test.js]: remove this line

File: src/tests/component/articles/Article.test.js [hunk 14]
```diff
@@ -0,0 +1,72 @@
...
+  addTag: jest.fn(),
+  removeTag: jest.fn()
+};
+// @spy function
+const spyFunc = (component, func) => {
+  const spy = jest.spyOn(component, func);
+  component.forceUpdate();
+  return spy;
+};
+describe('Article', () => {
```
Comment (nakiwuge) [inline src/tests/component/articles/Article.test.js]: add a new line

File: src/tests/component/articles/Article.test.js [hunk 15]
```diff
@@ -0,0 +1,72 @@
...
+  const component = shallow(<CreateArticle {...props} />);
+  test('should render without crashing', () => {
+    expect(component).toMatchSnapshot();
+  });
+  test('should create new article when user click button', () => {
+    const spy = spyFunc(component.instance(), 'onSubmit');
+    const fakeEvent = { preventDefault: () => {} };
+    const form = component.find('[data-test="G-submitData"]');
```
Comment (nakiwuge) [inline src/tests/component/articles/Article.test.js]: add new line

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**